### DISCUSSION TITLE
Issue with Tooling Setup instructions

### DISCUSSION THREAD

**Discussion by marmitegeek** [twbs/bootstrap#31582] [General]
The [instructions for setting up Bootstrap's build tools](https://getbootstrap.com/docs/4.5/getting-started/build-tools/#tooling-setup) do not work for me.
>3. Install Ruby, install Bundler with `gem install bundler`, and finally run `bundle install`. This will install all Ruby dependencies, such as Jekyll and plugins.

After running `bundle install` I get...

~~>username@pcname:~/bootstrap$ bundle install~~
~>Could not locate Gemfile~

~It seems that after running `npm install` in step 2, the Gemfile (among others) gets deleted.~
~This is assuming you downloaded the [source code](https://getbootstrap.com/docs/4.5/getting-started/download/#source-files) from the site manually, rather than via a package manager - When I do `npm install bootstrap`, I never get the Gemfile.~
~I'm assuming I'm not meant to get the Gemfile if doing it via npm, in which case this should be documented IMO.~

~After re-adding all the files that npm deleted, `bundle install` works...until it doesn't...~
>username@pcname:~/bootstrap$ bundle install
>Fetching gem metadata from https://rubygems.org/.........
>Using public_suffix 4.0.5
>Following files may not be writable, so sudo is needed:
>  /usr/local/bin
>  /var/lib/gems/2.7.0
>  /var/lib/gems/2.7.0/build_info
>  /var/lib/gems/2.7.0/cache
>  /var/lib/gems/2.7.0/doc
>  /var/lib/gems/2.7.0/extensions
>  /var/lib/gems/2.7.0/gems
>  /var/lib/gems/2.7.0/specifications
>Using addressable 2.7.0
>Using bundler 2.1.4
>Using colorator 1.1.0
>Using concurrent-ruby 1.1.6
>Fetching eventmachine 1.2.7
>
>
>Your user account isn't allowed to install to the system RubyGems.
>  You can cancel this installation and run:
>
>      bundle install --path vendor/bundle
>
>  to install the gems into ./vendor/bundle/, or you can enter your password and install the bundled gems to RubyGems using sudo.
>
>  Password:
>Installing eventmachine 1.2.7 with native extensions
>Gem::Ext::BuildError: ERROR: Failed to build gem native extension.
>
>    current directory: /tmp/bundler20200902-2133-elykdneventmachine-1.2.7/gems/eventmachine-1.2.7/ext/usr/bin/ruby2.7 -I /usr/lib/ruby/2.7.0 -r ./siteconf20200902-2133-1rr3prs.rb extconf.rb
>checking for -lcrypto... *** extconf.rb failed ***
>Could not create Makefile due to some reason, probably lack of necessary libraries and/or headers.  Check the mkmf.log file for more details.  You may need configuration options.
>
>Provided configuration options:
>        --with-opt-dir
>        --without-opt-dir
>        --with-opt-include
>        --without-opt-include=${opt-dir}/include
>        --with-opt-lib
>        --without-opt-lib=${opt-dir}/lib
>        --with-make-prog
>        --without-make-prog
>        --srcdir=.
>        --curdir
>        --ruby=/usr/bin/$(RUBY_BASE_NAME)2.7
>        --with-ssl-dir
>        --without-ssl-dir
>        --with-ssl-include
>        --without-ssl-include=${ssl-dir}/include
>        --with-ssl-lib
>        --without-ssl-lib=${ssl-dir}/lib
>        --with-openssl-config
>        --without-openssl-config
>        --with-pkg-config
>        --without-pkg-config
>        --with-crypto-dir
>        --without-crypto-dir
>        --with-crypto-include
>        --without-crypto-include=${crypto-dir}/include
>        --with-crypto-lib
>        --without-crypto-lib=${crypto-dir}/lib
>        --with-cryptolib
>        --without-cryptolib
>/usr/lib/ruby/2.7.0/mkmf.rb:471:in `try_do': The compiler failed to generate an executable file. (RuntimeError)
>You have to install development tools first.
>        from /usr/lib/ruby/2.7.0/mkmf.rb:564:in `try_link0'
>        from /usr/lib/ruby/2.7.0/mkmf.rb:582:in `try_link'
>        from /usr/lib/ruby/2.7.0/mkmf.rb:801:in `try_func'
>        from /usr/lib/ruby/2.7.0/mkmf.rb:1029:in `block in have_library'
>        from /usr/lib/ruby/2.7.0/mkmf.rb:971:in `block in checking_for'
>        from /usr/lib/ruby/2.7.0/mkmf.rb:361:in `block (2 levels) in postpone'
>        from /usr/lib/ruby/2.7.0/mkmf.rb:331:in `open'
>        from /usr/lib/ruby/2.7.0/mkmf.rb:361:in `block in postpone'
>        from /usr/lib/ruby/2.7.0/mkmf.rb:331:in `open'
>        from /usr/lib/ruby/2.7.0/mkmf.rb:357:in `postpone'
>        from /usr/lib/ruby/2.7.0/mkmf.rb:970:in `checking_for'
>        from /usr/lib/ruby/2.7.0/mkmf.rb:1024:in `have_library'
>        from extconf.rb:8:in `block in check_libs'
>        from extconf.rb:8:in `all?'
>        from extconf.rb:8:in `check_libs'
>        from extconf.rb:95:in `<main>'
>
>To see why this extension failed to compile, please check the mkmf.log which can be found here:
>
>  /tmp/bundler20200902-2133-elykdneventmachine-1.2.7/extensions/x86_64-linux/2.7.0/eventmachine-1.2.7/mkmf.log
>
>extconf failed, exit code 1
>
>Gem files will remain installed in /tmp/bundler20200902-2133-elykdneventmachine-1.2.7/gems/eventmachine-1.2.7 for inspection.
>Results logged to
>/tmp/bundler20200902-2133-elykdneventmachine-1.2.7/extensions/x86_64-linux/2.7.0/eventmachine-1.2.7/gem_make.out
>
>An error occurred while installing eventmachine (1.2.7), and Bundler cannot continue.
>Make sure that `gem install eventmachine -v '1.2.7' --source 'https://rubygems.org/'` succeeds before bundling.
>
>In Gemfile:
>  jekyll-redirect-from was resolved to 0.16.0, which depends on
>    jekyll was resolved to 4.1.1, which depends on
>      em-websocket was resolved to 0.5.1, which depends on
>        eventmachine

After this I'm lost. Until today, I'd never used Ruby in my life.
Anyone any suggestions on where to go from here?

**Comment by ffoodd** [comment]
~~NPM only delivers Bootstrap itself, not the website sources. You need to clone Bootstrap's repository through Git to get the Gemfile—and you're right, that's missing as a second step in our doc.~~

Forget it, I misread. 

First of all, you shouldn't run `npm install bootstrap`, only `npm install`. This is what confused this.
Next, it seems you're not admin: did you entered your password to use `sudo`?

**Reply by marmitegeek** [reply]
Hmmm... I think I messed up somewhere and ended up with a github clone and the npm version - Confused myself with missing Gemfiles.
I've stricken that part form my point.
I think it's worth mentioning that the doc's step-by-step assumes you already have a copy of the source (instead of the compiled version) before starting. This, perhaps, isn't as clear as it could be.

When running bundle as root, you're explicitly told not to do it...
> Don't run Bundler as root. Bundler can ask for sudo if it is needed, and installing your bundle as root will break this application for all non-root users on this machine.

In my paste, I am asked for a password shortly before the error.
I thought, at first, I simply mistyped my password, but that does not seem to be the case.
From what I can tell, the error is one of the following...

- something wrong with the Gemfile
- something wrong with "eventmachine"
- a lack of understanding on my part

**Reply by ffoodd** [reply]
I agree with the lack of clarity about having sources (either by downloading or by using `git clone`)— I first thought that was your main issue.

Since our CI builds correctly, Gemfile is good, as well as `eventmachine`. 

Bundler recommends not using `sudo`, albeit your log is clear: 
> Your user account isn't allowed to install to the system RubyGems.

So I guess you don't have much choice.

I'm not sure what to recommend to help you with this, for now :/
